src/core/geometry/qgspoint.cpp

```diff
@@ -674,7 +684,7 @@ QgsPoint QgsPoint::project( double distance, double azimuth, double inclination
 
 bool QgsPoint::isEmpty() const
 {
-  return false;
+  return std::isnan( mX ) || std::isnan( mY );
```

Comment:
QgsGeometry::asPoint SIP code will need updating to raise an exception if an empty point is attempted to convert to a QgsPointXY

Reply:
@nyalldawson Why?
An empty QgsPoint could be converted as an empty QgsPointXY, https://github.com/qgis/QGIS/pull/9645/files#diff-79b56e84ace3386453d7ffde919a808bR35